File: frontend/mobile/apple/iOS/My Home Agent/My Home Agent/Auth/AuthManager.swift
import Foundation
import LocalAuthentication

enum AuthError: Error, LocalizedError {
    case noRefreshToken
    case refreshFailed
    case loginFailed
    case networkError(NetworkError)
    case credentialsNotFound
    case biometricError(String)
    case biometricsFailed
    case noSessionId
    case sessionExpired
    
    public var errorDescription: String? {
        switch self {
        case .noRefreshToken:
            return "No refresh token available. Please log in again."
        case .refreshFailed:
            return "Failed to refresh authentication. Please log in again."
        case .loginFailed:
            return "Login failed. Please check your credentials and try again."
        case .networkError(let error):
            return error.errorDescription
        case .credentialsNotFound:
            return "No stored credentials found. Please log in with your email and password."
        case .biometricError(let message):
            return "Biometric authentication failed: \(message)"
        case .biometricsFailed:
            return "Biometric authentication failed. Please try again or use your password."
        case .noSessionId:
            return "No session ID found. Please log in again."
        case .sessionExpired:
            return "Your session has expired. Please log in again."
        }    
    }
}

final class AuthManager {
    static let shared = AuthManager()
    
    private let keychainService = "com.hestami.app"
    private let accessTokenKey = "accessToken"
    private let refreshTokenKey = "refreshToken"
    private let sessionIdKey = "sessionId"
    private let emailKey = "userEmail"
    private let passwordKey = "userPassword"
    private let rememberMeKey = "rememberMe"
    // User ID key removed
    
    private init() {}
    
    // Access token getter/setter with secure Keychain storage
    var accessToken: String? {
        get {
            return KeychainWrapper.standard.string(forKey: accessTokenKey)
        }
        set {
            if let newValue = newValue {
                _ = KeychainWrapper.standard.set(newValue, forKey: accessTokenKey)
            } else {
                _ = KeychainWrapper.standard.removeObject(forKey: accessTokenKey)
            }
        }
    }
    
    // Refresh token getter/setter with secure Keychain storage
    var refreshToken: String? {
        get {
            return KeychainWrapper.standard.string(forKey: refreshTokenKey)
        }
        set {
            if let newValue = newValue {
                _ = KeychainWrapper.standard.set(newValue, forKey: refreshTokenKey)
            } else {
                _ = KeychainWrapper.standard.removeObject(forKey: refreshTokenKey)
            }
        }
    }
    
    // Session ID getter/setter with secure Keychain storage
    var sessionId: String? {
        get {
            return KeychainWrapper.standard.string(forKey: sessionIdKey)
        }
        set {
            if let newValue = newValue {
                _ = KeychainWrapper.standard.set(newValue, forKey: sessionIdKey)
            } else {
                _ = KeychainWrapper.standard.removeObject(forKey: sessionIdKey)
            }
        }
    }
    
    // Stored email getter/setter
    var storedEmail: String? {
        get {
            return KeychainWrapper.standard.string(forKey: emailKey)
        }
        set {
            if let newValue = newValue {
                _ = KeychainWrapper.standard.set(newValue, forKey: emailKey)
            } else {
                _ = KeychainWrapper.standard.removeObject(forKey: emailKey)
            }
        }
    }
    
    // Stored password getter/setter (only when Remember Me is enabled)
    var storedPassword: String? {
        get {
            return isRememberMeEnabled ? KeychainWrapper.standard.string(forKey: passwordKey) : nil
        }
        set {
            if let newValue = newValue, isRememberMeEnabled {
                _ = KeychainWrapper.standard.set(newValue, forKey: passwordKey)
            } else {
                _ = KeychainWrapper.standard.removeObject(forKey: passwordKey)
            }
        }
    }
    
    // Remember Me setting
    var isRememberMeEnabled: Bool {
        get {
            return UserDefaults.standard.bool(forKey: rememberMeKey)
        }
        set {
            UserDefaults.standard.set(newValue, forKey: rememberMeKey)
            if !newValue {
                // Clear stored password if Remember Me is disabled
                _ = KeychainWrapper.standard.removeObject(forKey: passwordKey)
            }
        }
    }
    
    // User ID property removed
    
    var isAuthenticated: Bool {
        // Consider the user authenticated if they have session cookies
        return NetworkManager.shared.hasSessionCookie()
    }
    
    // Validate if the current session is active by attempting to fetch the user profile
    func validateSession() async -> Bool {
        // First check if we have a session cookie
        guard NetworkManager.shared.hasSessionCookie() else {
            AppLogger.auth.debug("No session cookie found, session is invalid")
            return false
        }
        
        // If we have a cookie, verify it works by fetching the profile
        do {
            AppLogger.auth.debug("Validating session by fetching user profile")
            let _: User = try await NetworkManager.shared.request(
                endpoint: "/api/users/profile/", 
                method: .get,
                parameters: [:] // Empty parameters for POST request
            )
            AppLogger.auth.info("Session is valid")
            return true
        } catch {
            AppLogger.auth.warning("Session validation failed: \(error.localizedDescription, privacy: .public)")
            return false
        }
    }
    
    // Verify authentication setup by checking cookies and headers
    func verifyAuthenticationSetup() {
        #if DEBUG
        AppLogger.auth.debug("===== AUTHENTICATION VERIFICATION SUMMARY =====")
        AppLogger.auth.debug("isAuthenticated = \(self.isAuthenticated, privacy: .public)")
        
        // Use NetworkManager's cookie debugging
        NetworkManager.shared.debugCookies()
        
        AppLogger.auth.debug("===== END AUTHENTICATION VERIFICATION =====")
        #endif
    }
    
    // Login method with email and password
    func login(email: String, password: String, rememberMe: Bool = false) async throws -> User {
        AppLogger.debugSensitive("Attempting login with email", sensitiveData: email, category: AppLogger.auth)
        let parameters = ["email": email, "password": password]
        
        AppLogger.auth.debug("Sending login request to /api/users/login/")
        
        do {
            let loginResponse: LoginResponse = try await NetworkManager.shared.request(
                endpoint: "/api/users/login/",
                method: .post,
                parameters: parameters
            )
            
            AppLogger.auth.info("Login successful")
            
            // Store credentials if Remember Me is enabled
            self.isRememberMeEnabled = rememberMe
            self.storedEmail = email
            if rememberMe {
                self.storedPassword = password
            } else {
                self.storedPassword = nil
            }
            
            // Verify that we have a session cookie after login
            if !NetworkManager.shared.hasSessionCookie() {
                AppLogger.auth.error("Warning - No session cookie found after login")
                AppLogger.auth.error("Login failed - server did not provide a session cookie")
                
                // Debug cookies after failed login
                NetworkManager.shared.debugCookies()
                
                throw AuthError.loginFailed
            }
            
            // Debug cookies after successful login
            NetworkManager.shared.debugCookies()
            
            return loginResponse.user
        } catch let networkError as NetworkError {
            // Convert NetworkError to AuthError with better message
            AppLogger.error("Login failed due to network error", error: networkError, category: AppLogger.auth)
            throw AuthError.networkError(networkError)
        } catch {
            // Handle other errors
            AppLogger.error("Login failed", error: error, category: AppLogger.auth)
            throw AuthError.loginFailed
        }
    }
    
    // Biometric authentication
    func authenticateWithBiometrics() async throws -> User {
        guard let email = storedEmail, let password = storedPassword else {
            throw AuthError.credentialsNotFound
        }
        
        // Use the main login method to ensure proper session validation
        return try await login(email: email, password: password, rememberMe: true)
    }
    
    // Refresh token method - No longer used as server doesn't provide tokens
    func refreshAccessToken() async throws {
        // Since the server doesn't use tokens, we'll just re-authenticate the user
        guard let email = storedEmail, let password = storedPassword else {
            throw AuthError.refreshFailed
        }
        
        AppLogger.auth.debug("Refreshing authentication")
        
        // Re-login with stored credentials
        _ = try await login(email: email, password: password, rememberMe: isRememberMeEnabled)
        // Note: login method already refreshes cookies
    }
    
    // Logout method
    func logout() {
        // Keep email for convenience but remove password
        if !isRememberMeEnabled {
            storedPassword = nil
        }
        
        // Authentication is now based solely on cookies
        AppLogger.auth.info("Logging out")
        
        // Clear all cookies
        NetworkManager.shared.clearCookies()
    }
    
    // Check if biometric authentication is available
    func canUseBiometricAuthentication() -> Bool {
        let context = LAContext()
        var error: NSError?
        
        // Check if the device supports biometric authentication
        let canEvaluate = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
        
        // Only return true if biometrics are available AND we have stored credentials
        return canEvaluate && storedEmail != nil && storedPassword != nil
    }
    
    // Register method for new user signup
    func register(
        email: String,
        password: String,
        confirmPassword: String,
        firstName: String,
        lastName: String,
        phoneNumber: String,
        userRole: String
    ) async throws -> User {
        AppLogger.debugSensitive("Attempting registration with email", sensitiveData: email, category: AppLogger.auth)
        
        let parameters: [String: Any] = [
            "email": email,
            "password": password,
            "confirm_password": confirmPassword,
            "first_name": firstName,
            "last_name": lastName,
            "phone_number": phoneNumber,
            "user_role": userRole,
            "service_provider": NSNull()
        ]
        
        AppLogger.auth.debug("Sending registration request to /api/users/register/")
        
        do {
            let registerResponse: RegisterResponse = try await NetworkManager.shared.request(
                endpoint: "/api/users/register/",
                method: .post,
                parameters: parameters
            )
            
            AppLogger.auth.info("Registration successful")
            return registerResponse.user
        } catch let networkError as NetworkError {
            AppLogger.error("Registration failed with network error", error: networkError, category: AppLogger.auth)
            throw AuthError.networkError(networkError)
        } catch {
            AppLogger.error("Registration failed", error: error, category: AppLogger.auth)
            throw AuthError.loginFailed
        }
    }
}

// Auth response models - Deprecated, use LoginResponse instead
struct AuthResponse: Decodable {
    let user: User
}

// RefreshResponse is no longer used as server doesn't provide tokens
struct RefreshResponse: Decodable {
    // Empty placeholder structure
}

// Registration response model
struct RegisterResponse: Decodable {
    let user: User
    let message: String?
}

File: frontend/mobile/google/android/MyHomeAgent/app/src/main/java/com/hestami_ai/myhomeagent/auth/AuthManager.kt
package com.hestami_ai.myhomeagent.auth

import android.content.Context
import android.content.SharedPreferences
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.core.content.edit
import androidx.fragment.app.FragmentActivity
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.hestami_ai.myhomeagent.data.model.User
import com.hestami_ai.myhomeagent.data.network.ApiService
import com.hestami_ai.myhomeagent.data.network.LoginRequest
import com.hestami_ai.myhomeagent.data.network.NetworkModule
import com.hestami_ai.myhomeagent.data.network.RegisterRequest
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import timber.log.Timber
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Authentication manager handling login, logout, and credential storage.
 * Matches iOS AuthManager.swift implementation.
 */
class AuthManager private constructor(context: Context) {

    companion object {
        private const val PREFS_NAME = "auth_prefs"
        private const val KEY_EMAIL = "userEmail"
        private const val KEY_PASSWORD = "userPassword"
        private const val KEY_REMEMBER_ME = "rememberMe"

        @Volatile
        private var instance: AuthManager? = null

        fun getInstance(context: Context): AuthManager {
            return instance ?: synchronized(this) {
                instance ?: AuthManager(context.applicationContext).also { instance = it }
            }
        }
    }

    private val encryptedPrefs: SharedPreferences
    private val regularPrefs: SharedPreferences
    private val apiService: ApiService

    init {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()

        encryptedPrefs = EncryptedSharedPreferences.create(
            context,
            PREFS_NAME,
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
        )

        regularPrefs = context.getSharedPreferences("${PREFS_NAME}_regular", Context.MODE_PRIVATE)
        apiService = NetworkModule.provideApiService()
    }

    // ==================== Stored Credentials ====================

    var storedEmail: String?
        get() = encryptedPrefs.getString(KEY_EMAIL, null)
        set(value) {
            encryptedPrefs.edit {
                if (value != null) putString(KEY_EMAIL, value)
                else remove(KEY_EMAIL)
            }
        }

    var storedPassword: String?
        get() = if (isRememberMeEnabled) encryptedPrefs.getString(KEY_PASSWORD, null) else null
        set(value) {
            encryptedPrefs.edit {
                if (value != null && isRememberMeEnabled) putString(KEY_PASSWORD, value)
                else remove(KEY_PASSWORD)
            }
        }

    var isRememberMeEnabled: Boolean
        get() = regularPrefs.getBoolean(KEY_REMEMBER_ME, false)
        set(value) {
            regularPrefs.edit { putBoolean(KEY_REMEMBER_ME, value) }
            if (!value) {
                // Clear stored password if Remember Me is disabled
                encryptedPrefs.edit { remove(KEY_PASSWORD) }
            }
        }

    // ==================== Authentication State ====================

    /**
     * Check if user is currently authenticated (has valid session cookie).
     */
    val isAuthenticated: Boolean
        get() = NetworkModule.hasActiveSession()

    /**
     * Validate if the current session is active by fetching user profile.
     */
    suspend fun validateSession(): Boolean = withContext(Dispatchers.IO) {
        if (!NetworkModule.hasActiveSession()) {
            Timber.d("No session cookie found, session is invalid")
            return@withContext false
        }

        try {
            Timber.d("Validating session by fetching user profile")
            val response = apiService.getProfile()
            val isValid = response.isSuccessful
            Timber.i(if (isValid) "Session is valid" else "Session validation failed")
            isValid
        } catch (e: Exception) {
            Timber.w(e, "Session validation failed")
            false
        }
    }

    // ==================== Login ====================

    /**
     * Login with email and password.
     */
    suspend fun login(
        email: String,
        password: String,
        rememberMe: Boolean = false
    ): User = withContext(Dispatchers.IO) {
        Timber.d("Attempting login with email: ${email.take(3)}***")

        try {
            val response = apiService.login(LoginRequest(email, password))

            if (!response.isSuccessful) {
                Timber.e("Login failed with status: ${response.code()}")
                throw AuthError.LoginFailed
            }

            val loginResponse = response.body() ?: throw AuthError.LoginFailed
            Timber.i("Login successful for user: %s", loginResponse.user.email)

            // Store credentials if Remember Me is enabled
            isRememberMeEnabled = rememberMe
            storedEmail = email
            storedPassword = if (rememberMe) password else null

            // Verify session cookie was set
            if (!NetworkModule.hasActiveSession()) {
                Timber.e("Warning - No session cookie found after login")
                throw AuthError.LoginFailed
            }

            loginResponse.user
        } catch (e: AuthError) {
            throw e
        } catch (e: Exception) {
            Timber.e(e, "Login failed")
            throw AuthError.NetworkError(e)
        }
    }

    // ==================== Registration ====================

    /**
     * Register a new user.
     */
    suspend fun register(
        email: String,
        password: String,
        confirmPassword: String,
        firstName: String,
        lastName: String,
        phoneNumber: String,
        userRole: String
    ): User = withContext(Dispatchers.IO) {
        Timber.d("Attempting registration with email: ${email.take(3)}***")

        try {
            val request = RegisterRequest(
                email = email,
                password = password,
                confirmPassword = confirmPassword,
                firstName = firstName,
                lastName = lastName,
                phoneNumber = phoneNumber,
                userRole = userRole
            )

            val response = apiService.register(request)

            if (!response.isSuccessful) {
                Timber.e("Registration failed with status: ${response.code()}")
                throw AuthError.RegistrationFailed
            }

            val registerResponse = response.body() ?: throw AuthError.RegistrationFailed
            Timber.i("Registration successful for user: %s", registerResponse.user.email)

            registerResponse.user
        } catch (e: AuthError) {
            throw e
        } catch (e: Exception) {
            Timber.e(e, "Registration failed")
            throw AuthError.NetworkError(e)
        }
    }

    // ==================== Biometric Authentication ====================

    /**
     * Check if biometric authentication is available and credentials are stored.
     */
    fun canUseBiometricAuthentication(context: Context): Boolean {
        val biometricManager = BiometricManager.from(context)
        val canAuthenticate = biometricManager.canAuthenticate(
            BiometricManager.Authenticators.BIOMETRIC_STRONG or
            BiometricManager.Authenticators.BIOMETRIC_WEAK
        ) == BiometricManager.BIOMETRIC_SUCCESS

        return canAuthenticate && storedEmail != null && storedPassword != null
    }

    /**
     * Authenticate using biometrics and stored credentials.
     */
    suspend fun authenticateWithBiometrics(activity: FragmentActivity): User {
        val email = storedEmail ?: throw AuthError.CredentialsNotFound
        val password = storedPassword ?: throw AuthError.CredentialsNotFound

        // Show biometric prompt
        val authenticated = showBiometricPrompt(activity)
        if (!authenticated) {
            throw AuthError.BiometricsFailed
        }

        // Use stored credentials to login
        return login(email, password, rememberMe = true)
    }

    private suspend fun showBiometricPrompt(activity: FragmentActivity): Boolean =
        suspendCancellableCoroutine { continuation ->
            val executor = ContextCompat.getMainExecutor(activity)

            val callback = object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    continuation.resume(true)
                }

                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    if (continuation.isActive) {
                        continuation.resumeWithException(
                            AuthError.BiometricError(errString.toString())
                        )
                    }
                }

                override fun onAuthenticationFailed() {
                    // Don't resume here - user can retry
                    Timber.d("Biometric authentication attempt failed")
                }
            }

            val promptInfo = BiometricPrompt.PromptInfo.Builder()
                .setTitle("Authenticate")
                .setSubtitle("Use your fingerprint or face to sign in")
                .setNegativeButtonText("Cancel")
                .build()

            val biometricPrompt = BiometricPrompt(activity, executor, callback)
            biometricPrompt.authenticate(promptInfo)

            continuation.invokeOnCancellation {
                biometricPrompt.cancelAuthentication()
            }
        }

    // ==================== Logout ====================

    /**
     * Logout and clear session.
     */
    fun logout() {
        Timber.i("Logging out")

        // Keep email for convenience but remove password if Remember Me is disabled
        if (!isRememberMeEnabled) {
            storedPassword = null
        }

        // Clear session cookies
        NetworkModule.clearSession()
        Timber.i("User logged out")
    }

    // ==================== Token Refresh ====================

    /**
     * Refresh authentication using stored credentials.
     * Note: Server uses session cookies, not tokens.
     */
    @Suppress("unused")
    suspend fun refreshAuthentication() {
        val email = storedEmail ?: throw AuthError.RefreshFailed
        val password = storedPassword ?: throw AuthError.RefreshFailed

        Timber.d("Refreshing authentication")
        login(email, password, rememberMe = isRememberMeEnabled)
    }
}
